packages/styletron-react/src/types.ts

```diff
@@ -0,0 +1,96 @@
+import * as React from "react";
+import type {ComponentType} from "react";
+import type {StyleObject} from "styletron-standard";
+
+export type AssignmentCommutativeReducerContainer = {
+  assignmentCommutative: true;
+  reducer: (a: StyleObject) => StyleObject;
+  style: StyleObject;
+  factory: (a: StyleObject) => AssignmentCommutativeReducerContainer;
+};
+
+export type NonAssignmentCommutativeReducerContainer = {
+  assignmentCommutative: false;
+  reducer: (b: StyleObject, a: any) => StyleObject;
+};
+
+export type ReducerContainer =
+  | AssignmentCommutativeReducerContainer
+  | NonAssignmentCommutativeReducerContainer;
+
+// TODO: more precise types
+export type Styletron = {
+  reducers: Array<ReducerContainer>;
+  base: any;
+  driver: any;
+  name?: string;
+  wrapper: any;
+  getInitialStyle: any;
+  ext?: {
+    name?: string;
+    base: any;
+    getInitialStyle: any;
+    with: any;
+  };
+  debug?: {
+    stackIndex: number;
+    stackInfo: {
+      stack: any;
+      stacktrace: any;
+      message: any;
+    };
+  };
+};
+
+export type StyletronProps<Props = {}> = Partial<{
+  $style: StyleObject | ((props: Props) => StyleObject);
+  $as: ComponentType<any> | keyof JSX.IntrinsicElements;
+  className: string;
+  /** @deprecated Use ref instead. Ref forwarding is enabled. */
+  $ref: Props extends {ref?: infer T} ? T : React.Ref<any>;
+  ref: Props extends {ref?: infer T} ? T : React.Ref<any>;
+}>;
+
+export type StyletronComponent<Props> = React.FC<
+  Props & StyletronProps<Props>
+> & {
+  __STYLETRON__: any;
+};
+
+export type StyledFn = {
```

Comment:
I wonder if typescript's function overloading would be more appropriate https://www.typescriptlang.org/docs/handbook/declaration-files/by-example.html#overloaded-functions and https://www.typescriptlang.org/docs/handbook/functions.html#overloads